libparsec/crates/client/src/workspace/transactions/file_operations.rs

```diff
@@ -272,122 +267,127 @@ fn prepare_truncate(
-        let chunks_skip_drain_until = last_block_last_chunk_index + 1;
-        // ...and finally we have an updated last block to append (containing the chunks ignored in the drain).
-        let maybe_new_last_block_chunks = Some(new_last_block_chunks);
-        (
-            blocks_drain_at,
-            chunks_skip_drain_until,
-            maybe_new_last_block_chunks,
-        )
-    };
+        // Those new chunks should not be removed
+        for chunk in &new_chunks {
+            removed_ids.remove(&chunk.id);
+        }
 
-    // Update the manifest with the new last block
+        // Truncate and add the new chunks
+        manifest.blocks.truncate(block);
+        manifest.blocks.push(new_chunks);
+    }
 
+    // Update the manifest
     manifest.need_sync = true;
     manifest.size = size;
     manifest.updated = timestamp;
-    // Only keep the blocks that should still be used as-is...
-    let removed_chunks = manifest
-        .blocks
-        .drain(blocks_drain_at..)
-        .flatten()
-        .skip(chunks_skip_drain_until)
-        .map(|x| x.id)
-        .collect();
-    // ...then re-insert the last block if only part of it is still used
-    if let Some(new_last_block_chunks) = maybe_new_last_block_chunks {
-        manifest.blocks.push(new_last_block_chunks);
-    }
 
-    removed_chunks
+    removed_ids
 }
 
 /// Prepare a resize operation by updating the provided manifest.
 ///
-/// Return a `Vec` of write operations that must be performed in order for the updated manifest to become valid.
-/// Each write operation consists of a new chunk to store, along with an offset to apply to the corresponding raw data.
-/// Note that the raw data also needs to be sliced to the chunk size and padded with null bytes if necessary.
-/// Also return a list of chunk IDs that must cleaned up from the storage, after the updated manifest has been successfully stored.
-pub(crate) fn prepare_resize(
+/// Return a `HashSet` of chunk IDs that must cleaned up from the storage, after the updated manifest has been successfully stored.
+pub fn prepare_resize(
     manifest: &mut LocalFileManifest,
     size: u64,
     timestamp: DateTime,
-) -> (Vec<WriteOperation>, Vec<ChunkID>) {
+) -> HashSet<ChunkID> {
     if size >= manifest.size {
         // Extend
-        prepare_write(manifest, 0, size, timestamp)
+        manifest.need_sync = true;
+        manifest.size = size;
+        manifest.updated = timestamp;
+        HashSet::new()
     } else {
         // Truncate
-        let removed_ids = prepare_truncate(manifest, size, timestamp);
-        (vec![], removed_ids)
+        prepare_truncate(manifest, size, timestamp)
     }
 }
 
-// Prepare reshape
-
 pub(crate) struct ReshapeBlockOperation<'a> {
     manifest_chunks: &'a mut Vec<Chunk>,
     reshaped_chunk: Chunk,
 }
 
 impl ReshapeBlockOperation<'_> {
+    pub fn new(chunks: &mut Vec<Chunk>) -> Option<ReshapeBlockOperation> {
```

Comment:
Rename `new` as `try_reshape` ? (Given `new` is expected to resurn `Self`, having a method with `try` or `maybe` emphasizes the Option in the return)